phantazm-zombies/src/main/java/com/github/phantazmnetwork/zombies/equipment/gun/GunState.java

```diff
@@ -0,0 +1,97 @@
+package com.github.phantazmnetwork.zombies.equipment.gun;
+
+import org.jetbrains.annotations.NotNull;
+
+public record GunState(long ticksSinceLastShot,
+                       long ticksSinceLastReload,
+                       int ammo,
+                       int clip,
+                       boolean isMainEquipment,
+                       int queuedShots) {
+
+    public @NotNull GunState.Builder toBuilder() {
+        return new Builder()
+                .setTicksSinceLastShot(ticksSinceLastShot)
+                .setTicksSinceLastReload(ticksSinceLastReload)
+                .setAmmo(ammo)
+                .setClip(clip)
+                .setMainEquipment(isMainEquipment)
+                .setQueuedShots(queuedShots);
+    }
+
+    public static class Builder {
+
+        private long ticksSinceLastShot;
+
+        private long ticksSinceLastReload;
+
+        private int ammo;
+
+        private int clip;
+
+        private boolean isMainEquipment;
+
+        private int queuedShots;
+
+        public long getTicksSinceLastShot() {
+            return ticksSinceLastShot;
+        }
+
+        public @NotNull Builder setTicksSinceLastShot(long ticksSinceLastShot) {
+            this.ticksSinceLastShot = ticksSinceLastShot;
+            return this;
+        }
+
+        public long getTicksSinceLastReload() {
```

Comment:
general question of if using a builder here is a bad thing
originally I wanted GunState to be immutable when it is passed around in case something wanted to store it, but I don't know if trying to guarantee that is dumb and it would be better to only make it mutable by the Gun and nothing else

as for the comment, I did just to make sure that you can get data that you've set

Reply:
i don't think using a builder here is bad, and fair enough on the getters